vernac/declareDef.ml

```diff
@@ -14,6 +14,29 @@ open Entries
 open Globnames
 open Impargs
 
+(* Hooks naturally belong here as they apply to both definitions and lemmas *)
+module Hook = struct
+  module S = struct
+    type t = UState.t
+      -> (Names.Id.t * Constr.t) list
+      -> Decl_kinds.locality
+      -> Names.GlobRef.t
+      -> unit
+  end
+
+  type t = S.t CEphemeron.key
+
+  let make hook = CEphemeron.create hook
+
+  let call ?hook ?fix_exn uctx trans l c =
+    try Option.iter (fun hook -> CEphemeron.get hook uctx trans l c) hook
```

Comment:
I'd like to have a specific error if the hook is not there (eg the ephemeron was sent over a wire)

Reply:
Maybe I did not get what this PR does wrt the ephemeron stuff but I guess there are 3 cases:
- no hook is set
- hook is set but CEphemeron.get returns None
- hook is set and we get Some
Are the first two paths mixed? I believe that the second should be detected and properly reported.

Reply:
In the second case indeed the CEphemeron will raise `Invalid_key`, this is the standard reporting mechanism for "ephemerons", how would you like to be reported?

Note that this works with current clients as the new behavior is strictly a superset of the old one, before this PR you always get an invalid client, now you get it only if the hook was set.

Reply:
The API is a bit more convenient like this, but the CEphemeron could be indeed placed in the proof info field.

Reply:
> The API is a bit more convenient like this, but the CEphemeron could be indeed placed in the proof info field.

So the argument in favor of doing that is that now `DeclareDef` doesn't need to know about the Ephemeron, the argument against is that clients do need to be aware of the internal representation of the hook. I dunno.

Reply:
Note @gares  that indeed after this and #10316 clients would have better options w.r.t. analyzing what kind of proof they have on hand [for example the `respects opacity` thing is not longer needed], however so far the PRs keep proof information private on purpose, analysis functions should be added as clients need I guess. In any case decisions such as the exact location of Ephemerons should evolve with use cases and IMHO are not a big deal in the sense that types keep us safe when doing such refactorings.

Reply:
I was suggesting to catch the exception here and give an ad-hoc anomaly mentioning "proof terminator" (and possibly the constant name).

Reply:
Actually this doesn't to have to be fatal, clients may want to catch `InvalidKey` and do something else [like re-schedule the proof?]

Reply:
I don't think it can happen: declaredef is called *always* from the very same actor that started the definition (eg that passed the hook in the first place).

The hack in the STM that was putting hooks back was there because hooks were part of the proof state that is sent back and forth, and hence pruned wrt hooks. This should have gone now.

There was also another "exception" that I don't know if it is still there. The worker would "run Qed" not really to update the environment but to check the proof term is OK (free from tactic errors) before sending it back. In that case a dummy hook could be used (I don't recall how this is/was done). A more polite thing would be to call the kernel, but IIRC the API were not easy to call, so I just did as if Qed was run.

Reply:
> I don't think it can happen: declaredef is called _always_ from the very same actor that started the definition (eg that passed the hook in the first place).

I guess it'd depend what the client does with the proof information.

But indeed the STM should have a better time [specially after #10316 ] but I am not so familiar with that it does [at least now it does indeed not preserve the lemma info, but after #10316 as you say the same actor could keep a hold on it.]

Reply:
The Ephemeron is necessary yet by the way as the STM will try to marshall the full state for example on `par:` .